src/transport/raw/TCP.cpp

```diff
@@ -290,101 +268,82 @@ CHIP_ERROR TCPBase::SendAfterConnect(const PeerAddress & addr, System::PacketBuf
+            err = ProcessSingleMessage(peerAddress, buffer, messageSize);
             SuccessOrExit(err);
-
-            err = endPoint->AckReceive(messageSize);
-            SuccessOrExit(err);
-            continue;
-        }
-
-        // Buffer is incomplete if we reach this point
-        if (buffer->HasChainedBuffer())
-        {
-            buffer->CompactHead();
-            continue;
         }
-
-        if (messageSize > 0)
-        {
-            // Open the receive window just enough to allow the remainder of the message to be received.
-            // This is necessary in the case where the message size exceeds the TCP window size to ensure
-            // the peer has enough window to send us the entire message.
-            uint16_t neededLen = static_cast<uint16_t>(messageSize - buffer->DataLength());
-            err                = endPoint->AckReceive(neededLen);
-            SuccessOrExit(err);
-        }
-
-        // Buffer is incomplete and we cannot get more data
-        break;
     }
 
 exit:
+    // Ack the data consumed.
+    endPoint->AckReceive(consumed);
+
+    // Return any partial message to the endpoint, and wait for more.
     if (!buffer.IsNull())
     {
-        // Incomplete processing will be retried
         endPoint->PutBackReceivedData(std::move(buffer));
     }
 
     return err;
 }
 
+CHIP_ERROR TCPBase::ProcessSingleMessage(const PeerAddress & peerAddress, System::PacketBufferHandle & buffer, uint16_t messageSize)
+{
+    // We enter with `buffer` containing at least one full message, perhaps in a chain.
+    // `buffer->Start()` currently points to the message data.
+    // On exit, `buffer` will have had `messageSize` bytes consumed, no matter what.
+    System::PacketBufferHandle message;
+    if (buffer->DataLength() == messageSize)
+    {
+        // In this case, the head packet buffer contains exactly the message.
+        // This is common because typical messages fit in a network packet, and are delivered as such.
+        // Peel off the head to pass upstream.
+        message = buffer.Retain();
+        buffer.Advance();
+    }
+    else
+    {
+        // The message is either longer or shorter than the head buffer.
+        // In either case, copy the message to a fresh linear buffer to pass upstream.
+        message = System::PacketBufferHandle::New(messageSize, 0);
+        if (message.IsNull())
+        {
+            // Either we are out of memory, or the message is too large to allocate.
+            buffer.Consume(messageSize);
+            return CHIP_ERROR_NO_MEMORY;
+        }
+        buffer->Read(message->Start(), messageSize);
+        buffer.Consume(messageSize);
+        message->SetDataLength(messageSize);
+    }
+
+    PacketHeader header;
+    ReturnErrorOnFailure(header.DecodeAndConsume(message));
+
+    // message receive handler will attempt to free the buffer, however as the buffer may
+    // contain additional data, we retain it to prevent actual free
```

Comment:
This comment is no longer correct, and we should remove it.  We're just handing over `message` to the receive handler, and it can do whatever it wants with it.  We no longer care about `message` after this point.